security-framework/src/secure_transport.rs

```diff
@@ -699,31 +700,35 @@ impl SslContext {
         unsafe { cvt(SSLSetProtocolVersionMin(self.0, min_version.0)) }
     }
 
-    /// Returns the Application Layer Protocol Negotiation (ALPN) protocols list.
+    /// Returns the set of protocols selected via ALPN if it succeeded.
     #[cfg(feature = "OSX_10_13")]
     pub fn alpn_protocols(&self) -> Result<Vec<String>> {
-        let protocols: CFArray<CFString> = unsafe {
-            let mut protocols_ref = mem::zeroed();
-            cvt(SSLCopyALPNProtocols(self.0, &mut protocols_ref as *mut _))?;
-            CFArray::wrap_under_create_rule(protocols_ref)
-        };
+        unsafe {
+            let mut array = ptr::null();
+            cvt(SSLCopyALPNProtocols(self.0, &mut array))?;
+            if array.is_null() {
+                return Ok(vec![]);
+            }
 
-        Ok(protocols
-            .into_iter()
-            .map(|proto| proto.to_string())
-            .collect())
+            let array = CFArray::<CFString>::wrap_under_create_rule(array);
+            Ok(array.into_iter().map(|p| p.to_string()).collect())
+        }
     }
 
-    /// Sets the Application Layer Protocol Negotiation (ALPN) protocols list.
+    /// Configures the set of protocols use for ALPN.
+    ///
+    /// This is only used for client-side connections.
     #[cfg(feature = "OSX_10_13")]
     pub fn set_alpn_protocols(&mut self, protocols: &[&str]) -> Result<()> {
```

Comment:
I'm wondering if it's better to change the signature of `protocols` to `&[String]`? This way we can avoid creating a temporary vector in `ClientBuilder`

Reply:
Yeah I was going back and forth with that. The nice part about `&[&str]` is that it's almost always going to be called with static string literals like `ctx.set_alpn_protocols(&["h2", "http/1.1"])` and it'd be nice to not have to do `ctx.set_alpn_protocols(&["h2".to_string(), "http/1.1".to_string()])`. We could just make it generic over `T: AsRef<str>`, but that's not really a pattern we've used elsewhere in the library.